### PR TITLE
fix: correct parsing issues with quoted strings

### PR DESCRIPTION
**Pull request by pgguru**
An initial stab at fixing the parsing issues seen in issue #7.

### CODE REVIEW

File: parseutils.c [hunk 1]
```diff
@@ -380,44 +383,97 @@ strip_quotes(char *source, char quote)
...
- * 
+ *   var="abc=123"
+ *   multiline="multi\nline"
+ *   quoted="{\"quoted\":\"json\"}"
+ *
  * Return two tokens; strip the quotes around the second one.
  * If exactly two tokens are not found, throw an error.
  */
 char **
 parse_keqv_line(char *line)
```
Comment (cbandy) [inline parseutils.c:506]: For reference, these strings are in Go's `%q` format. It looks like the decisions for how format non-printable characters and such are made in [strconv.Quote](https://pkg.go.dev/strconv#Quote).

- https://github.com/kubernetes/kubernetes/blob/v1.22.0/pkg/volume/downwardapi/downwardapi.go#L262
- https://github.com/kubernetes/kubernetes/blob/v1.22.0/pkg/fieldpath/fieldpath.go#L71
- https://github.com/kubernetes/kubernetes/blob/v1.22.0/pkg/fieldpath/fieldpath.go#L36
- https://pkg.go.dev/fmt
- https://pkg.go.dev/strconv#Quote

Comment (cbandy) [reply parseutils.c:506]: Except for the `\x` cases, the description looks very similar to JSON.

Comment (cbandy) [reply parseutils.c:506]: It is not JSON, of course. There are other escape sequences it emits. [This function](https://github.com/golang/go/blob/go1.17/src/strconv/quote.go#L68) includes `\a`, `\U`, `\v`, and `\x` among others.

Comment (pgguru) [reply parseutils.c:506]: Cool. Thanks for the source; now I can write a real parser for it! :)

Comment (pgguru) [reply parseutils.c:506]: Parser written to handle this now.

**Comment by pgguru** [comment]
The string parser has been brought up-to-date with Go's particular method of quoting.  Example of how things work now from the output with synthetic data (i.e., not directly in an annotated pod, but using the same file that you would):

image

**Comment by pgguru** [comment]
Note: when merging, we will want to merge and squash.

**Comment by pgguru** [comment]
For reference on parsing spec:

https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/strconv/quote.go;drc=refs%2Ftags%2Fgo1.17.3;l=31

File: parseutils.c [hunk 2]
```diff
@@ -355,69 +358,178 @@ parse_ss_line(char *line, int *ntok)
...
+				*dst++ = '\v';
+				src++;
+				break;
+			case '"':
+				*dst++ = '"';
+				src++;
+				break;
+			case 'x':
+				/* next 2 chars are hex bytes */
+				if (is_hex_digit(src[1]) && is_hex_digit(src[2]))
```
Comment (cbandy) [inline parseutils.c:429]: Should this (and the 'u' 'U' loop below) be more defensive about the length of src? The last two characters the buffer being `\x`, for example.

**Comment by pgguru** [comment]
The `is_hex_digit()` check catches those cases; a nul byte fails the check,
so it'll hit the error path and bail.

File: parseutils.c (hunk 2 above)
Comment (pgguru) [reply parseutils.c:429]: The `is_hex_digit()` check catches those cases; a nul byte fails the check,
so it'll hit the error path and bail.

**Pull request merged**